src/liballoc/vec.rs

```diff
@@ -2016,25 +2032,214 @@ where
+where
+    I: Iterator<Item = T>,
+{
+    default fn from_iter(iterator: I) -> Self {
+        SpecFromNested::from_iter(iterator)
+    }
+}
+
+// A helper struct for in-place iteration that drops the destination slice of iteration.
+// The source slice is dropped by IntoIter
+struct InPlaceDrop<T> {
+    inner: *mut T,
+    dst: *mut T,
+}
+
+impl<T> InPlaceDrop<T> {
+    unsafe fn len(&self) -> usize {
+        self.dst.offset_from(self.inner) as usize
+    }
+}
+
+impl<T> Drop for InPlaceDrop<T> {
+    #[inline]
+    fn drop(&mut self) {
+        unsafe {
+            ptr::drop_in_place(slice::from_raw_parts_mut(self.inner, self.len()) as *mut _);
+        }
+    }
+}
+
+impl<T> SpecFrom<T, IntoIter<T>> for Vec<T> {
+    fn from_iter(iterator: IntoIter<T>) -> Self {
+        // A common case is passing a vector into a function which immediately
+        // re-collects into a vector. We can short circuit this if the IntoIter
+        // has not been advanced at all.
+        // We can also reuse the memory and move the data to the front if
+        // allocating a new vector and moving to it would result in the same capacity
+        let non_zero_offset = iterator.buf.as_ptr() as *const _ != iterator.ptr;
+        if !non_zero_offset || iterator.len() >= iterator.cap / 2 {
+            unsafe {
+                let iterator = mem::ManuallyDrop::new(iterator);
+                if non_zero_offset {
+                    ptr::copy(iterator.ptr, iterator.buf.as_ptr(), iterator.len());
+                }
+                let vec = Vec::from_raw_parts(iterator.buf.as_ptr(), iterator.len(), iterator.cap);
+                return vec;
+            }
+        }
+
+        let mut vec = Vec::new();
+        // must delegate to spec_extend() since extend() itself delegates
+        // to spec_from for empty Vecs
+        vec.spec_extend(iterator);
+        vec
+    }
+}
+
+// Further specialization potential once
+// https://github.com/rust-lang/rust/issues/62645 has been solved:
+// T can be split into IN and OUT which only need to have the same size and alignment
+impl<T, I> SpecFrom<T, I> for Vec<T>
+where
+    I: Iterator<Item = T> + InPlaceIterable + SourceIter<Source: AsIntoIter<T>>,
+{
+    default fn from_iter(mut iterator: I) -> Self {
+        // This specialization only makes sense if we're juggling real allocations.
+        // Additionally some of the pointer arithmetic would panic on ZSTs.
+        if mem::size_of::<T>() == 0 {
+            return SpecFromNested::from_iter(iterator);
+        }
+
+        let (src_buf, src_end) = {
+            let inner = unsafe { iterator.as_inner().as_into_iter() };
+            (inner.buf.as_ptr(), inner.end)
+        };
+        let dst = src_buf;
+
+        let dst = if mem::needs_drop::<T>() {
+            // special-case drop handling since it prevents vectorization
+            let mut sink = InPlaceDrop { inner: src_buf, dst };
+            let _ = iterator.try_for_each::<_, Result<_, !>>(|item| {
```

Comment:
If iteration panics it seems like we're going to double-drop all of the remaining elements in the Vec via InPlaceDrop.

Reply:
With in place iteration the vec acts as sink and source as the same time. `IntoIter` is the source and will drop the tail of the vec on panic. `InPlaceDrop` acts as sink and will drop the head. At least that's the intent. Did I mix something up?

Reply:
When initialized, the InPlaceDrop covers every element of the inner vector. If the try_for_each panics e.g. immediately before advancing at all, its drop impl will drop every element of the vector after InPlaceDrop does, right?

Reply:
I have added a test that should show that drop handling works as expected. `InPlaceDrop` drops the front, `IntoIter` drops the back.

Testing specialization is tricky though since there are no tools to ensure that you're actually executing the methods you want to test. Please check if the test satisfies your concern.